internal/pkg/schematic/schematic.go

```diff
@@ -0,0 +1,291 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+// Package schematic provides a service to translate schematic IDs to schematics.
+package schematic
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path/filepath"
+
+	"github.com/klauspost/compress/zstd"
+	"github.com/siderolabs/image-factory/pkg/schematic"
+	"github.com/siderolabs/talos/pkg/machinery/extensions"
+	"github.com/u-root/u-root/pkg/cpio"
+	"github.com/ulikunitz/xz"
+	"go.uber.org/zap"
+	"golang.org/x/sync/singleflight"
+	"gopkg.in/yaml.v3"
+)
+
+const (
+	// OfficialExtensionPrefix is the prefix for official extensions.
+	OfficialExtensionPrefix = "siderolabs/"
+
+	// ImageFactoryHost is the host for the Talos image factory.
+	ImageFactoryHost = "factory.talos.dev"
+)
+
+type Service struct {
+	logger   *zap.Logger
+	cacheDir string
+}
+
+func NewService(cacheDir string, logger *zap.Logger) (*Service, error) {
+	if cacheDir == "" {
+		userCacheDir, err := os.UserCacheDir()
+		if err != nil {
+			return nil, fmt.Errorf("failed to get user cache dir: %w", err)
+		}
+
+		cacheDir = filepath.Join(userCacheDir, "talemu-schematics")
+	}
+
+	if logger == nil {
+		logger = zap.NewNop()
+	}
+
+	return &Service{
+		cacheDir: cacheDir,
+		logger:   logger,
+	}, nil
+}
+
+func (svc *Service) GetByID(ctx context.Context, id string) (*schematic.Schematic, error) {
```

Comment:
The logic:
- if we already cached the schematic for the given id, read it from the filesystem.
- otherwise, download initramfs from the factory and extract this information from it, then put it to the cache.